refactor: `Query` classes and related code

DaniBodor opened the pull request:
In this PR:
- `Query` class has been converted into a dataclass and all common arguments are defined in the parent class instead of the children.
- `ProteinProteinInterfaceResidueQuery` and `ProteinProteinInterfaceAtomicQuery` were merged into a single child dataclass of `Query`, as were `SingleResidueVariantResidueQuery` and `SingleResidueVariantAtomicQuery`
  - `resolution` is now given as an attribute instead of being a separate class
- Moved as much code as possible into parent class
  - unbound functions were unified and made internal methods of parent.
  - unused/obsolete functions were removed:
    - code related to adding hydrogens
    - `_get_atom_node_key` function
    - potentially more that I've lost track of
  - build function is largely defined in parent class, with specifics defined in child class as internal `_build_helper` methods.
    - it feels like build function could be further unified, but my tests doing this kept failing.
- Reorganized module
  - `QueryCollection` at the bottom now
- Update external modules to match this
  - tests
  - docs
  - notebooks
  - etc

Not yet done:
- Unify `build_atomic_graph` and `build_residue_graph` from buildgraph.py (#506)
- Separate `radius` from `distance_cutoff` for PPI Queries (#504)


fixes: #480 and #490 

Final TO DO list:
- [x] Move parameters to parent class
- [x] Switch to dataclasses
- [x] Unify atomic and residue version of each query into one
    - [x] fix all tests, linting, READMEs, notebooks until here
- [x] Move as much as possible into parent class
    - [x] Make main build function in parent, have specific helper functions in child classes.
    - [x] Remove unbound functions or make them methods or parent if still useful.
    - [x] Unify build methods for PPI and SVR (if possible)
- [x] Make sure all tests and external modules work with new classes 
- [x] Update docstrings
    - [x] Check how this should work with inheritance, what is listed in which docstring, etc
    - [ ] Check if it's possible to inherit docstrings. 
        - Haven't found a nice way of doing this.
    - [x] Check for all methods as well (not all necessarily require one).
- [x] Remove code related to inlcuding hydrogens
    - [x] check that this is indeed OK
- [ ] Clarify `distance_cutoff` vs `radius` across different `Query` classes (see [here](https://github.com/DeepRank/deeprank2/issues/460#issuecomment-1713910900))
    - Moved to PR #504 instead
- [x] Resolve conflicts with #497 
    - [x] Potentially others
- [x] Check all #TODOs throughout script
- [ ] Refactor duplicate code from buildgraph module
    - Will be a new PR for issue #506

DaniBodor commented:
reopened as #486

DaniBodor commented:
How do you feel about current class names? They are a bit long.
Other options are 
1. `ProteinProteinInterfaceQuery` and `SingleResidueVariantQuery` (current)
2. `PPIQuery` and `SRVQuery` (starting capitals are a bit odd maybe)
3. `PpiQuery` and `SrvQuery` (not capitalizing it also looks odd to me)
4. `QueryPPI` and `QuerySRV`
5. other ideas?

Is `DeepRankQuery` too long as well for the parent class? Maybe I should've left it `Query`...

DaniBodor commented:
Also, happy to get feedback on any `TODO`s left in the code

DaniBodor commented:
I found a red flag, that sent me down a rabbit hole of red flags:

https://github.com/DeepRank/deeprank2/blob/13f50680a227eff30af65977c939ca28cf7f6517/deeprank2/query.py#L299-L311

variant residue is only added to atomic graphs, but not residue graphs
- adding the variant to a residue graph as well breaks the tests (not sure yet if it should)
- removing the variant from the atom graph DOES NOT break the tests (I think it probably should be tested somewhere)
  - while trying to add a test, I
    - realized that it's not obvious how to check whether a residue is part of a graph
    - realized that test_query.py::test_variant_graph_101M lists the wronf wildtype amino acid, and this didn't break the tests
    - realized that the chain_id gets mangled somewhere in the process and it's not obvious how and where and how to build a `Residue` object that matches the one in the graph
    - that the amino acid is not checked in the `__eq__` function of `Residue`

NB: all this was already true before refactoring. Compare:
   - Atomic query (line 628): https://github.com/DeepRank/deeprank2/blob/3203d7ca2bd8d06e21c255565d465a89233c7ec3/deeprank2/query.py#L623-L639
   - Residue query: https://github.com/DeepRank/deeprank2/blob/3203d7ca2bd8d06e21c255565d465a89233c7ec3/deeprank2/query.py#L472-L481

gcroci2 commented on deeprank2/query.py:
```diff
@@ -93,28 +95,45 @@ def _check_pssm(pdb_path: str, pssm_paths: Dict[str, str], suppress: bool, verbo
+    pdb_path: str
+    resolution: str
+    chain_ids: List[str] | str
+    pssm_paths: Dict[str, str] = field(default_factory=dict)
```
Just briefly, what's the need for using `field` here and why it's not used for `chain_ids` as well?

gcroci2 commented on deeprank2/query.py:
```diff
@@ -447,199 +452,52 @@ def build(self, feature_modules: List[ModuleType], include_hydrogens: bool = Fal
+                    for atom in residue.atoms:
+                        atoms.add(atom)
+            atoms = list(atoms)
+            #TODO: why was this a set at first? I think each atom is unique anyway, given that it has a Residue property
```
Your comment makes sense, but I guess there must be some reason why it could be that atoms are not unique. Otherwise indeed it could be a list from the beginning. What about asking Coos?

DaniBodor replied:
I tested whether we can make them a list right away, and the tests now fail. I don't really understand why, but I think for the sake of wrapping this up I would just leave it as is (I had already removed the #TODO in a later commit)

gcroci2 commented on deeprank2/utils/buildgraph.py:
```diff
@@ -318,4 +318,4 @@ def get_surrounding_residues(structure: Union[Chain, PDBStructure], residue: Res
             close_residues.add(structure_atom.residue)
 
-    return close_residues
+    return list(close_residues)
```
`get_surrounding_residues` misses the type hint about the returned variable

gcroci2 replied:
Returning a list instead of a set, in the atomic resolution build case may cause an issue if (and only if) `variant_residue` is not unique (is this a possible scenario?), and is appended to the list `residues` again. It's also true that `atoms` is then defined as a set and populated according to `residues`, so even if the latter scenario is possible, I guess that defining then `atoms` as a set could avoid the issue. Maybe you've already carefully thought about that and we've already reached a conclusion, but just in case here's my comment. Let me know if I need to double check this more carefully.

DaniBodor replied:
The way I remember it, but would need to double check, is that a set was returned here, which was immediately converted to a list by the method that uses this. I short-cut that by converting to list before returning.
The logic didn't change. atoms are added to a set, so that duplicates are culled. That set is then converted to a list for usage by the code.

DaniBodor replied:
also, see my comment above.

DaniBodor replied:
> `get_surrounding_residues` misses the type hint about the returned variable

fixed in f42752a4

gcroci2 reviewed (commented):
I'm not done yet, few comments in the meantime:

gcroci2 commented on deeprank2/query.py:
```diff
@@ -551,161 +551,30 @@ def _load_ppi_pssms(pssm_paths: Optional[Dict[str, str]],
-        self._pssm_paths = pssm_paths
+    def __post_init__(self):
+        super().__post_init__()
+        if len(self.chain_ids) != 2:
```
I think the previous logic is more flexible and we shouldn't change it, i.e., giving the possibility to explicitly indicate the two chains involved. Even if a structure has more than 2 chains, it should be possible to specify which are the chains to be considered for the interface.

DaniBodor replied:
I think, but again will need to check, that the `chain_ids` argument is used 
- by SRVs: to determine which chain the variant is on
- by PPIs: to determine which 2 chains form the interface
This does not prohibit other chains from being part of the graph. I will check the documentation and make sure that this is not misleading.

gcroci2 replied:
Makes sense, and now the docs are clearer :) I think I was a bit confused here but looking at it now seems good

gcroci2 commented on deeprank2/query.py:
```diff
@@ -716,40 +585,40 @@ def build(self, feature_modules: List[ModuleType], include_hydrogens: bool = Fal
+        if self.resolution == 'atomic':
+            graph = build_atomic_graph(contact_atoms, self.get_query_id(), self.distance_cutoff)
+        elif self.resolution == 'residue':
+            residues_selected = {atom.residue for atom in contact_atoms}
```
Instead of making `residues_selected` a list in the line below, you could do it directly here (residues_selected = list{atom.residue for atom in contact_atoms})

gcroci2 commented on deeprank2/query.py:
```diff
@@ -183,19 +184,30 @@ def build(
         Returns:
             :class:`Graph`: The resulting :class:`Graph` object with all the features and targets.
         """
+        # TODO: Should this be an internal method, as it is only called upon by QueryCollection and not by user?
```
Could be that a user is interested in building the graph only. I'd keep it as it is for now.

DaniBodor replied:
OK. removed the TODO in $$$$$

gcroci2 commented on deeprank2/query.py:
```diff
@@ -183,19 +184,30 @@ def build(
-            elif isinstance(feature_modules, str):
-                feature_modules = [importlib.import_module('deeprank2.features.' + feature_modules)]
+        # TODO: I would prefer if  `_set_feature_modules` would return a list of `ModuleType` instead of `str`,
+        # but this leads to an exception in the pool function in `QueryCollection.process`, that I don't know how to solve.
```
This seems more for a separate issue, and needs more details about the exception

gcroci2 commented on deeprank2/query.py:
```diff
@@ -183,19 +184,30 @@ def build(
         self._pssm_required = conservation in feature_modules
 
-        return self._child_build(feature_modules)
+        # TODO: my gut feeling is that the building can be unified further, but it is
```
I think it's fine; it's always a trade off and you've already improved it quite a lot.

DaniBodor replied:
Yes and no. I think as we start adding more query subclasses, it will be very helpful to have this as unified as possible. But I think we will run into that as we start adding more query types, and can deal with it then. As you said, it is now already much easier to add more classes and understand what the similarities and differences between the current ones are.

gcroci2 commented on tests/test_query.py:
```diff
@@ -373,15 +373,15 @@ def test_incorrect_pssm_order():
     # check that error is thrown for incorrect pssm
     with pytest.raises(ValueError):
-        _ = q.build(conservation)
+        _ = q.build([conservation])
```
I'd rather add a check in the build method in which if the module is not a list (because it's only one) then we put it in a list, similarly to what we do for many other parameters.

DaniBodor replied:
OK. done in 1dd44efc

gcroci2 reviewed (commented):
Greeeeat job :) lots of redundant code has been deleted, we have less and clearer query classes, and also type hinting updated. The majority of comments are minor, few of them regards "strategic" decisions that is better we agree on.

- I'd revert `DeepRankQuery` to `Query`. I think that `DeepRank` doesn't add anything (only makes the class name longer), differently from the DeeprankDataset class, in which the org name was added to distinguish it from the PyTorch Dataset class (from which DeeprankDataset inherits).
- For the rest of queries' classes names, I'd leave them as they are for now. We still have to decide which new classes we'll have to implement, and when we do that we can also confirm or change the already existing ones (which are already shorter than before ;)). 
- About the red flag, could it be related to returning a list from `get_surrounding_residues` instead of a set? Anyway, we need to look at it together, I'm a bit lost in all the edits.

DaniBodor commented on deeprank2/query.py:
```diff
@@ -173,10 +173,11 @@ def _load_structure(
             for chain in structure.chains:
-                if chain.id in pssm_paths:
-                    pssm_path = pssm_paths[chain.id]
+                chain: Chain
```
yeah, sometimes VSCode wasn't recognizing the type correctly. In these cases I added the type hint. 
I think that in the final code, I managed to get rid of (almost) all type hints in the middle of code, by fixing the way things were sent around in the various functions.

DaniBodor replied:
removed in c81e48d2

DaniBodor replied:
removed in c81e48d2

DaniBodor commented on deeprank2/query.py:
```diff
@@ -136,35 +136,35 @@ def __post_init__(self):
 
     def _load_structure(
-        self, pdb_path: str, pssm_paths: Optional[Dict[str, str]],
+        self,
```
I think because this wasn't a method inside the class, but a standalone function.

DaniBodor commented on deeprank2/query.py:
```diff
@@ -447,199 +452,52 @@ def build(self, feature_modules: List[ModuleType], include_hydrogens: bool = Fal
             load_pssms = conservation == feature_modules
-        structure = self._load_structure(self._pdb_path, self._pssm_paths, include_hydrogens, load_pssms)
+            feature_modules = [feature_modules]
+        structure: PDBStructure = self._load_structure(include_hydrogens, load_pssms)
```
similar to comment above.

DaniBodor commented on deeprank2/query.py:
```diff
@@ -447,199 +452,52 @@ def build(self, feature_modules: List[ModuleType], include_hydrogens: bool = Fal
         variant_residue = None
-        for residue in structure.get_chain(self._chain_id).residues:
+        for residue in structure.get_chain(self.variant_chain_id).residues:
+            residue: Residue
```
see above

DaniBodor commented:
> I found a red flag, that sent me down a rabbit hole of red flags.
> 
> The initial red flag is that the variant residue is only added to atomic graphs, but not residue graphs:
> 
> https://github.com/DeepRank/deeprank2/blob/13f50680a227eff30af65977c939ca28cf7f6517/deeprank2/query.py#L299-L311
> 
> This made me try the following which concerned me:
> 
>     * adding the variant to a residue graph as well breaks the tests (not sure yet if it should)
> 
>     * removing the variant from the atom graph DOES NOT break the tests (I think it probably should be tested somewhere)
>       
>       * while trying to add a test for this, I realized that
>         
>         * the chain_id gets mangled somewhere in the process and it's not obvious how and where and how to build a `Residue` object that matches the one in the graph
>         * it's not obvious how to check whether a residue is part of a graph
>         * test_query.py::test_variant_graph_101M lists the wronf wildtype amino acid, and this didn't break the tests
>         * amino acid is not checked in the `__eq__` function of `Residue`
> 
> 
> NB: all this was already true before refactoring. Compare:
> 
>     * Atomic query (line 628): https://github.com/DeepRank/deeprank2/blob/3203d7ca2bd8d06e21c255565d465a89233c7ec3/deeprank2/query.py#L623-L639
> 
>     * Residue query: https://github.com/DeepRank/deeprank2/blob/3203d7ca2bd8d06e21c255565d465a89233c7ec3/deeprank2/query.py#L472-L481

I opened #511 to deal with this outside of this PR

DaniBodor commented on deeprank2/query.py:
```diff
@@ -716,40 +585,40 @@ def build(self, feature_modules: List[ModuleType], include_hydrogens: bool = Fal
+            #TODO: check whether this works with a set instead of a list
+        else:
+            raise NotImplementedError(f"No function exists to build graphs with resolution of {self.resolution}.")
+        graph.center = np.mean([atom.position for atom in contact_atoms], axis=0)
```
somehow the line got added twice, though (probably while resolving some conflict). I've now removed it once in 9a539329

DaniBodor commented on deeprank2/query.py:
```diff
@@ -716,40 +585,40 @@ def build(self, feature_modules: List[ModuleType], include_hydrogens: bool = Fal
+        if self.resolution == 'atomic':
+            graph = build_atomic_graph(contact_atoms, self.get_query_id(), self.distance_cutoff)
+        elif self.resolution == 'residue':
+            residues_selected = {atom.residue for atom in contact_atoms}
```
done in $$$$

DaniBodor commented on deeprank2/query.py:
```diff
@@ -412,11 +412,15 @@ class SingleResidueVariantQuery(DeepRankQuery):
             raise ValueError(f"SingleResidueVariantQuery must contain exactly 1 chain_id, but {len(self.chain_ids)} were given.")
         self.variant_chain_id = self.chain_ids[0]
 
+        if not self.distance_cutoff:
```
agreed! done in $$$$ 6e62c170

DaniBodor commented on deeprank2/query.py:
```diff
@@ -198,15 +199,10 @@ def _get_atom_node_key(atom) -> str:
-        self.model_id = value
-
-    @property
-    def targets(self) -> Dict[str, float]:
```
we removed it because it is extremely clunky to use properties in dataclasses when they are input arguments. It can still be inspected, even if not a property. It's just less protected, but we agreed (after consulting with Cunliang) that this would be outweighed by the benefit of having more readable code using dataclasses without properties.

DaniBodor commented on tests/utils/test_grid.py:
```diff
@@ -1,101 +1,48 @@
-    assert np.all(np.abs(grid.zs - target_zs) < coord_error_margin), f"\n{grid.zs} != \n{target_zs}"
-
-
-def test_atomic_grid_orientation():
```
The test was identical to the one above, except that it uses atomic instead of residue grids. 
I refactored so that the above test loops through the two levels of detail.
I have updated the name of that test to reflect this in $$$$

DaniBodor commented on deeprank2/query.py:
```diff
@@ -407,6 +405,11 @@ def queries(self) -> list[DeepRankQuery]:
 
+    @property
+    def ids_count(self) -> list[DeepRankQuery]:
+        """The list of queries added to the collection."""
```
ok, removed in $$$$$

DaniBodor commented on deeprank2/query.py:
```diff
@@ -416,114 +419,92 @@ def __iter__(self) -> Iterator[DeepRankQuery]:
                     augmentation = Augmentation(axis, angle)
-                    graph.write_as_grid_to_hdf5(output_path, grid_settings, grid_map_method, augmentation)
-
-            return None
```
I am pretty sure that by definition there is no functional difference between explicitly returning `None` and not explicitly stating `return`. See [here](https://stackoverflow.com/a/15300671/5170442). 

I don't believe any of the reasons stated in the thread for explicitly returning applies here (the function never returns anything and we are not trying to stop a process, as this is returned outside the `if` and thus only once the `try` anyway has passed.

@cbaakman, could you confirm?

cbaakman replied:
The return statement has no purpose.

DaniBodor commented on deeprank2/query.py:
```diff
@@ -416,114 +419,92 @@ def __iter__(self) -> Iterator[DeepRankQuery]:
-                cpu_count = cpu_count_system
-        self.cpu_count = cpu_count
-        _log.info(f'\nNumber of CPUs for processing the queries set to: {self.cpu_count}.')
+        self._prefix = "processed-queries" if not prefix else re.sub('.hdf5$', '', prefix)  # scrape extension if present
```
This is regarding `prefix`, right? Not `cpu_count`?
I don't understand what you mean with this. (It might be easier to explain in person).

In any case, I made it an attribute, because it is called by an internal method of `QueryCollection` and thought it would be easier not to have pass a variable that is a fixed variable initialized in the class. But can be changed back if that is preferred.

gcroci2 replied:
Indeed let's look at it together

gcroci2 replied:
This was about `_prefix`, `_cpu_count`, `_grid_settings`, and the other internal attributes that are defined for the first time in the `process()` method. Before this PR, they were just local variables in the method, now they are internal attributes. It's not a good practice to define attributes for the first time in some method of a class that is not the init (see my comment above). I see why you did it, i.e., for readability reasons for the developers; now the `_process_one_query` method is much clearer indeed. I am not sure in this case which is the best thing to do. Maybe we can ask Lourens and/or Cunliang?

DaniBodor replied:
Wouldn't the right thing be to add them all to the init of `QueryCollection`, and initialize them at `None`? And then make a better docstring explaining what they are and that they will be set properly in process? Then we both have them readable as attributes of `QueryCollection`, but also have a clean `_process_one_query` method

gcroci2 replied:
This was indeed what I suggested with my very first comment, fine for me :)

DaniBodor commented on deeprank2/query.py:
```diff
@@ -416,114 +419,92 @@ def __iter__(self) -> Iterator[DeepRankQuery]:
 
-        with Pool(self.cpu_count) as pool:
+        if grid_augmentation_count < 0:
+            raise ValueError(f"`grid_augmentation_count` may not be negative, but was given as {grid_augmentation_count}")
```
either is fine imo. I changed it in $$$$$

DaniBodor commented:
> Greeeeat job :) lots of redundant code has been deleted, we have less and clearer query classes, and also type hinting updated. The majority of comments are minor, few of them regards "strategic" decisions that is better we agree on.

Thanks :) 
I have now addressed most of your comments. A few remain unresolved and open (with questions from me). These might be easier addressed sitting down together as I think they are quick fixes once we agree how to proceed, but easier to discuss in person than typing ;)


> 
>     * I'd revert `DeepRankQuery` to `Query`. I think that `DeepRank` doesn't add anything (only makes the class name longer), differently from the DeeprankDataset class, in which the org name was added to distinguish it from the PyTorch Dataset class (from which DeeprankDataset inherits).

Makes sense. Done in 36cac04

> 
>     * For the rest of queries' classes names, I'd leave them as they are for now. We still have to decide which new classes we'll have to implement, and when we do that we can also confirm or change the already existing ones (which are already shorter than before ;)).
👍 We can always improve further when there are more classes.
> 
>     * About the red flag, could it be related to returning a list from `get_surrounding_residues` instead of a set? Anyway, we need to look at it together, I'm a bit lost in all the edits.

I created a new issue (#511) for this, because it is actually not part of this refactoring, but rather a separate (pre-existing) issue.

DaniBodor commented:
For some reason, Codacy is failing again, and now it doesn't even show my why anymore on their website. Can we maybe get rid of codacy after all and use something else to give us a quality badge?

DaniBodor commented:
The build is currently failing. Looks like the newest versions of some pyg additions are giving conflicts again. Maybe we should go back to specifying precise versions every so often do an update of this?
I will retry in a day or so and see if then the issues are solved.

gcroci2 commented:
> For some reason, Codacy is failing again, and now it doesn't even show my why anymore on their website. Can we maybe get rid of codacy after all and use something else to give us a quality badge?

I see that it's not tracking the current branch. Maybe it's that the issue?

gcroci2 commented:
> The build is currently failing. Looks like the newest versions of some pyg additions are giving conflicts again. Maybe we should go back to specifying precise versions every so often do an update of this? I will retry in a day or so and see if then the issues are solved.

Indeed, I think it's torch-cluster; 1.6.2 (or 1.6.1) should work, we can fix the version for this package only for now; they released 1.6.3 5 days ago

DaniBodor commented:
@gcroci2 , had you reviewed the changes to this yet? I see you gave a thumbs up, but haven't approved the PR yet.

DaniBodor commented:
> > For some reason, Codacy is failing again, and now it doesn't even show my why anymore on their website. Can we maybe get rid of codacy after all and use something else to give us a quality badge?
> 
> I see that it's not tracking the current branch. Maybe it's that the issue?

I now tracked this branch and it's giving a warning about a piece of code that wasn't touched here. Honestly, I feel that this codacy thing is giving way more headaches than any actual benefit it is providing. Can we just get rid of it? Or at least turn it off as a check and only use it for quality measures?

gcroci2 commented:
> > > For some reason, Codacy is failing again, and now it doesn't even show my why anymore on their website. Can we maybe get rid of codacy after all and use something else to give us a quality badge?
> > 
> > 
> > I see that it's not tracking the current branch. Maybe it's that the issue?
> 
> I now tracked this branch and it's giving a [warning ](https://app.codacy.com/gh/DeepRank/deeprank2/pullRequest?prid=12567459)about a piece of code that wasn't touched here. Honestly, I feel that this codacy thing is giving way more headaches than any actual benefit it is providing. Can we just get rid of it? Or at least turn it off as a check and only use it for the code quality score?

Does it make sense to keep it only for the quality score? I mean, if everything is turned off then how is the score computed? In case, we can just eliminate the codacy thing entirely

DaniBodor commented:
I think it was only this comment left: https://github.com/DeepRank/deeprank2/pull/492#discussion_r1337159863

I now think you were talking about `cpu_count` and not `prefix`, right? I first thought `prefix`, because that was the last line tagged, but then your comment doesn't make sense to me. If it's about `cpu_count`, I can change it now.

DaniBodor commented:
> I think it was only this comment left: [#492 (comment)](https://github.com/DeepRank/deeprank2/pull/492#discussion_r1337159863)
> 
> I now think you were talking about `cpu_count` and not `prefix`, right? I first thought `prefix`, because that was the last line tagged, but then your comment doesn't make sense to me. If it's about `cpu_count`, I can change it now (done in 00f24b4).

I think then it's ready for review again.

gcroci2 reviewed (approved):
Approved, I only have another comment about the attributes thing (see [here](https://github.com/DeepRank/deeprank2/pull/492#discussion_r1377834541)). Let me know what you decide at the end, and if you ask somebody what he/she tells you.

Also, should we remove codacy with this PR?

DaniBodor commented:
I fixed the attribute issue above in 3cd2a8c.
Can you also take a look at the TODOs I put in lines 250 and 335 (same issue) and 462 and give me your opinion on those. When that is done, I can finally close this PR.

DaniBodor commented:
> Also, should we remove codacy with this PR?

no, let's remove it in a separate PR? This one is already so big and unwieldly, that I don't want to add anything that doesn't need to be here.

gcroci2 commented:
> > Also, should we remove codacy with this PR?
> 
> no, let's remove it in a separate PR? This one is already so big and unwieldly, that I don't want to add anything that doesn't need to be here.

Done in #517

gcroci2 commented:
> I fixed the attribute issue above in [3cd2a8c](https://github.com/DeepRank/deeprank2/commit/3cd2a8ce71358e807721f3f4a0fa56267e6c30b3). Can you take a look and see if this is what you were thinking?
> 
> Can you also take a look at the TODOs I put in lines 250 and 335 (same issue) and 462 of query.py and give me your opinion on those.
> 
> When that is done, I can finally merge this PR :)
> 
> Note:
> 
> * re the TODO in line 38: I will leave as is now and remove the #TODO
> * re the TODO in line 525: I made a new issue ([Make `MapMethod` an attribute of `GridSettings` in deeprank2.grid module #516](https://github.com/DeepRank/deeprank2/issues/516)) for this and will remove the #TODO

line 250: I think it's actually fine to throw an error since only 1 chain identifier is needed to identify the chain where the variant is. If more are given then the user hasn't understood the purpose of `chain_ids`.
line 335: same
line 462: I don't see any TODO. 

Also, if there are more things to discuss let's just have a meeting because finding stuff in this PR is becoming very time consuming and inefficient ;)

One final thing: looking at those lines made me think about something I commented earlier about, but was lost someway. I think it should be clearer that `chain_ids` refers to the chains involved in the definition of whatever query (e.g., the chain identifier of the variant), but does not limit the usage of other eventual chains present in the pdb files. It could be indeed that the radius or the cutoff distance defined is enough to include other chains, not considered in the `chain_ids` variable because its purpose is another one. So, for example in `SingleResidueVariantQuery`, the doc string says `the chain identifier(s) in the pdb file (generally a single capital letter)` but I think it should be something like `the chain identifier of the variant residue in the pdb file (generally a single capital letter)`. For `ProteinProteinInterfaceQuery`, it could be something like `the two chain identifiers of the interface in the pdb file (generally single capital letters)`.

DaniBodor commented:
> > I fixed the attribute issue above in [3cd2a8c](https://github.com/DeepRank/deeprank2/commit/3cd2a8ce71358e807721f3f4a0fa56267e6c30b3). Can you take a look and see if this is what you were thinking?

Have you looked at this also?


> > Can you also take a look at the TODOs I put in lines 250 and 335 (same issue) and 462 of query.py and give me your opinion on those.


> line 462: I don't see any TODO.

This is about error catching in `_process_one_query`:
Right now, if there's an error, it's caught and that query is skipped. My guess is that this was put in place so that it didnt crash due to a handfull of faulty pdb files.
My question was whether we want to make this optional, and by default raise an error. I think generally, users should have clean data to work with, and if not pro-actively set this to ignore errors. What do you think?


> One final thing: looking at those lines made me think about something I commented earlier about, but was lost someway. I think it should be clearer that `chain_ids` refers to the chains involved in the definition of whatever query (e.g., the chain identifier of the variant), but does not limit the usage of other eventual chains present in the pdb files. It could be indeed that the radius or the cutoff distance defined is enough to include other chains, not considered in the `chain_ids` variable because its purpose is another one. So, for example in `SingleResidueVariantQuery`, the doc string says `the chain identifier(s) in the pdb file (generally a single capital letter)` but I think it should be something like `the chain identifier of the variant residue in the pdb file (generally a single capital letter)`. For `ProteinProteinInterfaceQuery`, it could be something like `the two chain identifiers of the interface in the pdb file (generally single capital letters)`.
> 
Clarified in 23357f08
